File: src/main/java/com/enderryno/nuclearcraft/events/blocks/CustomBlockPlaceEvent.java
package com.enderryno.nuclearcraft.events.blocks;

import com.enderryno.nuclearcraft.NuclearCraft;
import com.enderryno.nuclearcraft.classes.CustomItem;
import com.enderryno.nuclearcraft.interfaces.PluginItem;
import com.enderryno.nuclearcraft.services.BlockRegister;
import com.enderryno.nuclearcraft.services.ItemRegister;
import org.bukkit.NamespacedKey;
import org.bukkit.block.Block;
import org.bukkit.block.TileState;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.block.BlockPlaceEvent;
import org.bukkit.inventory.ItemStack;
import org.bukkit.persistence.PersistentDataContainer;
import org.bukkit.persistence.PersistentDataType;

public class CustomBlockPlaceEvent implements Listener {

    @EventHandler
    public void onBlockPlace(BlockPlaceEvent event) {
        // Get the item in the player's hand
        ItemStack item = event.getItemInHand();
        PluginItem customItem = null;
        Block block = event.getBlock();

        // Get persistent data container
        PersistentDataContainer container = item.getItemMeta().getPersistentDataContainer();

        // Get key and check if namespace exists
        NamespacedKey blockIdKey = new NamespacedKey(NuclearCraft.getPlugin(NuclearCraft.class), "custom-block-id");
        NamespacedKey itemIdKey = new NamespacedKey(NuclearCraft.getPlugin(NuclearCraft.class), "item-id");
        if (!container.has(blockIdKey, PersistentDataType.STRING)) return;
        if (!container.has(itemIdKey, PersistentDataType.STRING)) return;

        // Get the custom block id and set it to the block
        String customBlockId = container.get(blockIdKey, PersistentDataType.STRING);
        String customItemId = container.get(itemIdKey, PersistentDataType.STRING);

        try {
             customItem = ItemRegister.getRegisteredItems().get(customItemId);
        } catch (Exception e) {
            e.printStackTrace();
        }

        try {
            BlockRegister.getRegisteredBlocks()
                    .get(customBlockId)
                    .placeBlock(customItem, block.getLocation());
        } catch (Exception e) {
            e.printStackTrace();
        }

    }
}

File: src/main/java/com/enderryno/nuclearcraft/events/blocks/CustomBlockPlaceEvent.kt
package com.enderryno.nuclearcraft.events.blocks

import com.enderryno.nuclearcraft.NuclearCraft
import com.enderryno.nuclearcraft.interfaces.PluginItem
import com.enderryno.nuclearcraft.services.BlockRegister
import com.enderryno.nuclearcraft.services.ItemRegister
import org.bukkit.NamespacedKey
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.block.BlockPlaceEvent
import org.bukkit.persistence.PersistentDataType
import org.bukkit.plugin.java.JavaPlugin

class CustomBlockPlaceEvent : Listener {
    @EventHandler
    fun onBlockPlace(event: BlockPlaceEvent) {
        // Get the item in the player's hand
        val item = event.itemInHand
        var customItem: PluginItem? = null
        val block = event.block

        // Get persistent data container
        val container = item.itemMeta.persistentDataContainer

        // Get key and check if namespace exists
        val blockIdKey = NamespacedKey(JavaPlugin.getPlugin(NuclearCraft::class.java), "custom-block-id")
        val itemIdKey = NamespacedKey(JavaPlugin.getPlugin(NuclearCraft::class.java), "item-id")
        if (!container.has(blockIdKey, PersistentDataType.STRING)) return
        if (!container.has(itemIdKey, PersistentDataType.STRING)) return

        // Get the custom block id and set it to the block
        val customBlockId = container.get(blockIdKey, PersistentDataType.STRING)?: return
        val customItemId = container.get(itemIdKey, PersistentDataType.STRING)?: return

        try {
            customItem = ItemRegister.getRegisteredItems()?.get(customItemId)
        } catch (e: Exception) {
            e.printStackTrace()
        }
        try {
            BlockRegister.getBlock(customBlockId)
                    ?.placeBlock(customItem, block.location)
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }
}
